File: linear/integration-test/runge_kutta/runge_kutta.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
##############################################################################
# (c) Crown copyright 2021 Met Office. All rights reserved.
# The file LICENCE, distributed with this code, contains details of the terms
# under which the code may be used.
##############################################################################
'''
Run the linear model integration tests for the runge-kutta configuration
'''

import os
import re
import sys


from testframework import LFRicLoggingTest, TestEngine, TestFailed


class TLTest(LFRicLoggingTest):
    '''
    Run the linear model integration tests
    '''

    def __init__(self, flag):
        self._flag = flag
        if 'MPIEXEC_BROKEN' in os.environ:
            TLTest.set_mpiexec_broken()
        super(TLTest, self).__init__([sys.argv[1],
                                      'runge_kutta_configuration.nml',
                                      'test_' + self._flag],
                                     processes=1,
                                     name='tl_test.Log')

    def test(self, return_code, out, err):
        '''
        Error messages if the test failed to run
        '''
        if return_code != 0:
            message = 'Test program failed with exit code: {code}'
            raise TestFailed(message.format(code=return_code),
                             stdout=out, stderr=err,
                             log=self.getLFRicLoggingLog())

        # "out" becomes self.getLFRicLoggingLog() when PE>1
        if not self.test_passed(out):
            message = 'Test {} failed'
            raise TestFailed(message.format(self._flag),
                             stdout=out, stderr=err,
                             log=self.getLFRicLoggingLog())

        return 'TL test : '+self._flag

    def test_passed(self, out):
        '''
        Examine the output to see if the validity test passed
        '''
        success = False
        pattern = re.compile(r'\s+test\s+.*?:\s*PASS\s*$')
        for line in out.split("\n"):
            match = pattern.search(line)
            if match:
                success = True
        return success


class tl_test_kinetic_energy_gradient(TLTest):
    '''
    Test the kinetic energy gradient kernel
    '''
    def __init__(self):
        flag = "kinetic_energy_gradient"
        super(tl_test_kinetic_energy_gradient, self).__init__(flag)


class tl_test_rk_alg(TLTest):
    '''
    Test the runge-kutta timestepping
    '''
    def __init__(self):
        flag = "rk_alg"
        super(tl_test_rk_alg, self).__init__(flag)

class tl_test_project_eos_pressure(TLTest):
    def __init__(self):
        flag = "project_eos_pressure"
        super(tl_test_project_eos_pressure, self).__init__(flag)

class tl_test_advect_density_field(TLTest):
    '''
    Test density advection
    '''
    def __init__(self):
        flag = "advect_density_field"
        super(tl_test_advect_density_field, self).__init__(flag)


class tl_test_advect_theta_field(TLTest):
    '''
    Test theta advection
    '''
    def __init__(self):
        flag = "advect_theta_field"
        super(tl_test_advect_theta_field, self).__init__(flag)


class tl_test_vorticity_advection(TLTest):
    '''
    Test the vorticity advection kernel
    '''
    def __init__(self):
        flag = "vorticity_advection"
        super(tl_test_vorticity_advection, self).__init__(flag)


class tl_test_pressure_gradient_bd(TLTest):
    '''
    Test the pressure_gradient bd kernel
    '''
    def __init__(self):
        flag = "pressure_gradient_bd"
        super(tl_test_pressure_gradient_bd, self).__init__(flag)


class tl_test_hydrostatic(TLTest):
    '''
    Test the hydrostatic kernel
    '''
    def __init__(self):
        flag = "hydrostatic"
        super(tl_test_hydrostatic, self).__init__(flag)


class tl_test_timesteps(TLTest):
    '''
    Test running over multiple timesteps
    '''
    def __init__(self):
        flag = "timesteps"
        super(tl_test_timesteps, self).__init__(flag)


if __name__ == '__main__':
    TestEngine.run( tl_test_kinetic_energy_gradient() )
    TestEngine.run( tl_test_advect_density_field() )
    TestEngine.run( tl_test_advect_theta_field() )
    TestEngine.run( tl_test_vorticity_advection() )
    TestEngine.run( tl_test_hydrostatic() )
    TestEngine.run( tl_test_pressure_gradient_bd() )
    TestEngine.run( tl_test_project_eos_pressure() )
    TestEngine.run( tl_test_rk_alg() )

File: linear/integration-test/runge_kutta/runge_kutta.f90
!-----------------------------------------------------------------------------
! (C) Crown copyright 2021 Met Office. All rights reserved.
! The file LICENCE, distributed with this code, contains details of the terms
! under which the code may be used.
!-----------------------------------------------------------------------------

!>@brief   The top level program for the tangent linear tests.
!>@details The default is to run all available tests - which
!!         test whether the linear code is tangent linear to the
!!         corresponding nonlinear code.
program runge_kutta

  use mod_wait,           only : init_wait
  use mpi_mod,            only : initialise_comm, &
                                 finalise_comm
  use xios,               only : xios_initialize
  use log_mod,            only : log_event,       &
                                 LOG_LEVEL_ERROR, &
                                 LOG_LEVEL_INFO
  use tl_test_driver_mod, only : initialise,                  &
                                 finalise,                    &
                                 run_timesteps,               &
                                 run_kinetic_energy_gradient, &
                                 run_advect_density_field,    &
                                 run_advect_theta_field,      &
                                 run_vorticity_advection,     &
                                 run_project_eos_pressure,    &
                                 run_hydrostatic,             &
                                 run_pressure_gradient_bd,    &
                                 run_rk_alg
  implicit none

  character(*), parameter :: xios_id = "linear"

  character(:), allocatable :: filename
  integer                   :: world_communicator = -999
  integer                   :: model_communicator = -999

  ! Variables used for parsing command line arguments
  integer :: length, status, nargs
  character(len=0) :: dummy
  character(len=:), allocatable :: program_name, test_flag

  ! Flags which determine the tests that will be carried out
  logical :: do_test_timesteps = .false.
  logical :: do_test_kinetic_energy_gradient = .false.
  logical :: do_test_advect_density_field = .false.
  logical :: do_test_advect_theta_field = .false.
  logical :: do_test_project_eos_pressure = .false.
  logical :: do_test_vorticity_advection = .false.
  logical :: do_test_pressure_gradient_bd = .false.
  logical :: do_test_hydrostatic = .false.
  logical :: do_test_rk_alg = .false.

  ! Usage message to print
  character(len=256) :: usage_message

  ! Initialse mpi and create the default communicator: mpi_comm_world
  call initialise_comm( world_communicator )

  ! Initialise XIOS and get back the split mpi communicator. This requires
  ! an iodef.xml file to be available (even if use_xios_io is .false.).
  call init_wait()
  call xios_initialize(xios_id, return_comm = model_communicator)

  call log_event( 'TL testing running ...', LOG_LEVEL_INFO )

  ! Parse command line parameters
  call get_command_argument( 0, dummy, length, status )
  allocate(character(length)::program_name)
  call get_command_argument( 0, program_name, length, status )
  nargs = command_argument_count()

  ! Print out usage message if wrong number of arguments is specified
  if (nargs /= 2) then
     write(usage_message,*) "Usage: ",trim(program_name), &
          " <namelist filename> "      // &
          " test_XXX with XXX in { "   // &
          " timesteps, "               // &
          " kinetic_energy_gradient, " // &
          " advect_density_field, "    // &
          " advect_theta_field, "      // &
          " vorticity_advection, "     // &
          " pressure_gradient_bd, "    // &
          " project_eos_pressure, "    // &
          " hydrostatic, "             // &
          " rk_alg, "                  // &
          " } "
     call log_event( trim(usage_message), LOG_LEVEL_ERROR )
  end if

  call get_command_argument( 1, dummy, length, status )
  allocate( character(length) :: filename )
  call get_command_argument( 1, filename, length, status )

  call get_command_argument( 2, dummy, length, status )
  allocate(character(length)::test_flag)
  call get_command_argument( 2, test_flag, length, status )

  ! Choose test case depending on flag provided in the first command
  ! line argument
  select case (trim(test_flag))
  case ("test_timesteps")
     do_test_timesteps = .true.
  case ("test_kinetic_energy_gradient")
     do_test_kinetic_energy_gradient = .true.
  case ("test_advect_density_field")
     do_test_advect_density_field = .true.
  case ("test_advect_theta_field")
     do_test_advect_theta_field = .true.
  case ("test_project_eos_pressure")
     do_test_project_eos_pressure = .true.
  case ("test_vorticity_advection")
     do_test_vorticity_advection = .true.
  case ("test_pressure_gradient_bd")
     do_test_pressure_gradient_bd = .true.
  case ("test_hydrostatic")
     do_test_hydrostatic = .true.
  case ("test_rk_alg")
     do_test_rk_alg = .true.
  case default
     call log_event( "Unknown test", LOG_LEVEL_ERROR )
  end select

  call initialise( filename, model_communicator )
  deallocate( filename )

  if (do_test_timesteps) then
    call run_timesteps()
  endif
  if (do_test_kinetic_energy_gradient) then
    call run_kinetic_energy_gradient()
  endif
  if (do_test_advect_density_field) then
    call run_advect_density_field()
  endif
  if (do_test_advect_theta_field) then
    call run_advect_theta_field()
  endif
  if (do_test_vorticity_advection) then
    call run_vorticity_advection()
  endif
  if (do_test_project_eos_pressure) then
    call run_project_eos_pressure()
  endif
  if (do_test_pressure_gradient_bd) then
    call run_pressure_gradient_bd()
  endif
  if (do_test_hydrostatic) then
    call run_hydrostatic()
  endif
  if (do_test_rk_alg) then
    call run_rk_alg()
  endif

  call finalise()

  ! Finalise mpi and release the communicator
  call finalise_comm()

end program runge_kutta
